add documentation for ACI support (#1434)
radius-project/docs: docs/content/guides/author-apps/azure/howto-azure-container-instances/index.md

weight: 500
slug: 'azure-container-instances'
categories: "How-To"
tags: ["Azure","containers"]
---

This how-to guide will provide an overview of how to:

- Configure and deploy a Radius [Environment]({{< ref "/guides/deploy-apps/environments/overview" >}}) that uses [Azure Container Instances (ACI)](https://learn.microsoft.com/en-us/azure/container-instances/) as the compute provider.
- Define and deploy the demo application to the Environment using Radius to provision the necessary resources to run the application containers in ACI.

## Prerequisites

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
- [Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})
- Radius installed and initiated on a [supported Kubernetes cluster]({{< ref "/guides/operations/kubernetes/overview#supported-clusters" >}})
- Azure provider configured and registered with your Radius control plane, either through [Service Principal](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-sp/) or [Workload Identity](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-wi/)

> - A [user-assigned managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp) that has been [assigned](https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments-portal-managed-identity) to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription or resource group where the ACI will be deployed

Reviewer: Let me double check this but I think a user assigned identity needs to have subscription access for one of the API calls. I think system assigned doesnt need additional setup since it gets created with the NGroup at deploy time

Author: done.

{{< rad file="snippets/app.bicep" embed=true marker="//ENVIRONMENT">}}

> Note: be sure to replace the `resourceGroup` and `scope` values with your resource group ID and the `managedIdentity` value with your managed identity ID.

## Step 2: Deploy the Environment

1. Run the following command to deploy the Environment:
   ```bash
   rad deploy ./app.bicep --resource-group aciGroup
   ```

   You should see the following terminal output:

   ```
   Deployment In Progress...

   Deployment Complete

   Resources:
   aci-demo Applications.Core/environments
   ```

## Step 3: Define the Application and its resources

Add the application definition, along with Redis cache and gateway resources to the `app.bicep` file.

{{< rad file="snippets/app.bicep" embed=true marker="//APPLICATION" >}}

Add the application Container resource to the `app.bicep` file.

{{< rad file="snippets/app.bicep" embed=true marker="//CONTAINER" >}}

> > Notice that for ACI containers, you define a Gateway resource that provides L7 traffic for the container. Radius will provision the Gateway in Azure on your behalf and configure the container to use the Gateway as its ingress. The Gateway will be provisioned with a public IP address and a DNS name that you can use to access the application.

Reviewer: We should add that the alternative is an internal load balancer is used for just L4 traffic. An ILB is the default setup right now but if a gateway is provided, we'll use the gateway as the network profile for ACI instead

Reviewer: For the samples app a gateway is needed since we want to be able to connect to the demo site

Author: I think I'll leave out the alternative for the moment to avoid confusing the user too much in this how-to guide.

> ---

Reviewer 2: Is this a temporary placement under Azure? 

IMO this should be under containers or more broadly under a group that defines the core-resource types and have content that shows portability across runtimes

Also this can be figured on later and is not a blocker for merge.

Author: While I agree with you, I struggled to find an appropriate home for this within the current documentation structure. I think for now this is the best place for it, and in the future we can decide to move it once we have added the much needed extra documentation explaining cross-platform extensibility, etc.

type: docs
title: "How-To: Deploy an Application to Azure Container Instances"
linkTitle: "Deploy to ACI"
description: "Learn how to configure and deploy an application to Azure Container Instances"
weight: 500
slug: 'azure-container-instances'
categories: "How-To"
tags: ["Azure","containers"]
---

This how-to guide will provide an overview of how to:

- Configure and deploy a Radius [Environment]({{< ref "/guides/deploy-apps/environments/overview" >}}) that uses [Azure Container Instances (ACI)](https://learn.microsoft.com/en-us/azure/container-instances/) as the compute provider.

> - Define and deploy the demo application to the Environment using Radius, which provisions the necessary resources to run the application containers in ACI.

Reviewer 2: ```suggestion
- Define and deploy the demo application to the ACI Radius Environment, which provisions the necessary resources to run the application in ACI.
```

## Prerequisites

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})

> - [Bicep VSCode extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}})

Reviewer 2: ```suggestion
The [Bicep extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}}) for VS Code is recommended for Bicep language support
```

- Radius [installed]({{< ref "/guides/operations/kubernetes/kubernetes-install" >}}) on a [supported Kubernetes cluster]({{< ref "/guides/operations/kubernetes/overview#supported-kubernetes-clusters" >}})

> - An Azure provider configured and registered with your Radius control plane, either through [Service Principal](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-sp/) or [Workload Identity](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-wi/) that have been assigned to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription or resource group where the ACI containers will be deployed

Reviewer 2: should the Azure provider pre-requisite be after managed identity so its logically ordered?

Author: It doesn't really matter in terms of ordering, but I think it's appropriate to have them set up radius creds first.

Reviewer 2: The current cloud provider docs doesn't have details around using a existing identity to configure the provider. So it might be better for the user to set up the identity first and then configure it in Radius

- A [managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/) is [required]({{< ref "/reference/resource-schema/core-schema/environment-schema#identity" >}}) for ACI deployments, if you choose to utilize a [user-assigned managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp) then you need to ensure it is assigned to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription or resource group where the ACI containers will be deployed

## Step 1: Create a Radius Resource Group and Workspace

> Since the Radius control plane is hosted on your Kubernetes cluster, you'll need to create a new Radius Resource Group and Workspace so that you may target your application deployments to an ACI Environment. These will then be associated with the ACI Environment that you will configure and create in subsequent steps. 

Reviewer 2: ```suggestion
Since the Radius control plane is hosted on your Kubernetes cluster, you'll need to create a new Radius Resource Group and Workspace so that you can target your application deployments to an ACI Environment. These will then be associated with the ACI Environment that you will configure and create in subsequent steps. 
```

> Since the Radius control plane is hosted on your Kubernetes cluster, you'll need to create a new Radius Resource Group and Workspace so that you may target your application deployments to an ACI Environment. These will then be associated with the ACI Environment that you will configure and create in subsequent steps. 

Reviewer 2: qq for clarity: This step of creating resource grp and workspace is required even for kubernetes deployments right? just that its done by `rad init` as shown in most of our guides

Author: correct, but because we don't allow for `rad init` to initialize an environment external to the cluster on which Radius is hosted, we include these extra steps to set up the "external" environment (aka ACI)

> 1. Create a new Radius Resource Group called `aciGroup`:

Reviewer: Let's switch these two steps. So create the workspace first, then the resource group. We also need to add a `rad group switch aciGroup` so that the resource group gets added to the config.yaml under the workspace created

Author: done

   ```bash
   rad group create aciGroup
   ```

1. Then, create a new Radius Workspace called `aci-workspace`:
   ```bash
   rad workspace create kubernetes aci-workspace
   ```

> 

Reviewer: We also want to create the environment and switch to the environment similar to what is done for the resource group. This is so that we can update the config.yaml and ensure that the environment is tied to the correct workspace (note that nothing is really deployed when we do this, it's just for config purposes). So: 

`rad env create aci-demo -w aci-workspace` then `rad env switch aci-demo`

Author: done

## Step 2: Define the Environment with ACI compute

> Create a new file named `app.bicep` and add the following Environment definition. This Environment uses the `aci` compute provider and a user-assigned managed identity to provision the necessary resources for ACI and also registers a default Recipe for provisioning an Azure Redis Cache.

Reviewer 2: should the env definition to be in a separate file `env.bicep` so that there is separation as what goes into a app vs environment?

Author: good suggestion, updated.

{{< rad file="snippets/app.bicep" embed=true marker="//ENVIRONMENT">}}

> Note: be sure to replace the `resourceGroup` and `scope` values with your resource group ID and the `managedIdentity` value with your managed identity resource ID.

## Step 3: Deploy the Environment

1. Run the following command to deploy the Environment and associate it with the `aci-workspace` you created in the previous step:
   ```bash

>    rad deploy ./app.bicep --workspace aci-workspace

Reviewer: I think this needs to be --group instead of workspace

   ```

   You should see the following terminal output:

   ```
   Deployment In Progress...

   Deployment Complete

   Resources:
   aci-demo Applications.Core/environments
   ```

<br>

> Navigate to your resource group in the [Azure portal](https://portal.azure.com/) and you should see the relevant Azure resources that were provisioned by Radius for your ACI Environment, including the virtual network, internal load balancer, and network security group:

Reviewer 2: ```suggestion
Navigate to your resource group in the [Azure portal](https://portal.azure.com/) and you should see the relevant Azure resources that were provisioned by Radius for your ACI Radius Environment, including the virtual network, internal load balancer, and network security group:
```

{{< image src="azure-portal-env.png" alt="Screenshot of the Azure portal showing the resource group with the virtual network, internal load balancer, and network security group resources created by Radius" width=700px >}}

## Step 4: Define the Application and its resources

Add the application definition, along with Redis cache and gateway resources to the `app.bicep` file.

{{< rad file="snippets/app.bicep" embed=true marker="//APPLICATION" >}}

Add the application Container resource to the `app.bicep` file.

{{< rad file="snippets/app.bicep" embed=true marker="//CONTAINER" >}}

> 

Reviewer 2: Showing the final app.bicep file would be helpful here

Author: good point, I've simplified it to have the reader add the entire app.bicep in one single go so that they may have a view into the full file.

> Notice that for ACI containers, you define a Gateway resource that provides L7 traffic for the container. Radius will provision an Azure Application Gateway on your behalf and configure the container to use the Gateway as its ingress. The Gateway will be provisioned with a public IP address and a DNS name that you can use to access the application.

## Step 5: Deploy the Application

Run the following command to deploy the application:

```bash

> rad deploy ./app.bicep --workspace aci-workspace

Reviewer: I think this needs to be --group instead of workspace

```

> > Note that you are deploying the application specifically targeting the `aci-workspace` you had created in a previous step, which ensures that your application gets deployed to the ACI Environment. The same application can also be targeted to deploy into a Kubernetes Environment instead.

Reviewer 2: ```suggestion
> Note that you are deploying the application specifically targeting the `aci-workspace` you had created in a previous step, which ensures that your application gets deployed to the ACI Environment. The same application can also be targeted to deploy into a Kubernetes Radius Environment instead.
```

- Define and deploy the demo application to the ACI Radius Environment, which provisions the necessary resources to run the application in ACI.

## Prerequisites

- [rad CLI]({{< ref "installation#step-1-install-the-rad-cli" >}})
The [Bicep extension]({{< ref "installation#step-2-install-the-vs-code-extension" >}}) for VS Code is recommended for Bicep language support
- Radius [installed]({{< ref "/guides/operations/kubernetes/kubernetes-install" >}}) on a [supported Kubernetes cluster]({{< ref "/guides/operations/kubernetes/overview#supported-kubernetes-clusters" >}})
- An Azure provider configured and registered with your Radius control plane, either through [Service Principal](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-sp/) or [Workload Identity](https://docs.radapp.io/guides/operations/providers/azure-provider/howto-azure-provider-wi/) that have been assigned to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription or resource group where the ACI containers will be deployed

> - A [managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/) is [required]({{< ref "/reference/resource-schema/core-schema/environment-schema#identity" >}}) for ACI deployments, if you choose to utilize a [user-assigned managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp) then you need to ensure it is assigned to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription or resource group where the ACI containers will be deployed

Reviewer: ```suggestion
- A [managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/) is [required]({{< ref "/reference/resource-schema/core-schema/environment-schema#identity" >}}) for ACI deployments, if you choose to utilize a [user-assigned managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp) then you need to ensure it is assigned to the `Contributor` and `Azure Container Instances Contributor` roles on the subscription and resource group where the ACI containers will be deployed
```

We need to have subscription level access for some NGroup API calls with and user assigned identity

> ## Step 1: Create a Radius Resource Group and Workspace

Reviewer 2: ```suggestion
## Step 1: Create a Radius Resource Group, Workspace and a Radius Environment
```
